### ISSUE TITLE
Prefix icon gets lost

### ISSUE THREAD

**Issue by somospocos** [tobi-wan-kenobi/bumblebee-status#643]
#### How to reproduce

For both setups I'm using the Nerd-patched `Iosevka` font.

Contents of `~/.config/bumblebee-status/themes/icons/nerd-fonts.json`

    {
      "defaults": {
        "separator": "",
        "padding": "",
        "unknown": { "prefix": "" }
      },
      "osname": { "prefix": "" }
    }

Setup 1:

revision: legacy

i3 bar config:

    bar {
            id top
            workspace_buttons no
            font pango:Iosevka 9
            position top
            status_command python3 bumblebee-status/bumblebee-status \
                -m spacer:osname \
                -p osname.text="64 14.2" \
                --markup=pango \
                --iconmarkup="<span font_desc='Iosevka 15' rise='-3500'>{}</span>" \
                -t mustang-powerline \
                -i nerd-fonts
    }

legacy theme in `~/.config/bumblebee-status/themes/mustang-powerline.json`:

    {
        "icons": [ "awesome-fonts" ],
        "defaults": {
            "default-separators": false,
            "separator-block-width": 0,
            "prefixfg": "#ff8700"
        },
        "cycle": [
            {
                "fg": "#7e8aa2",
                "bg": "#202020"
            },
            {
                "fg": "#b1d631",
                "bg": "#2d2d2d"
            }
        ]
    }

How does it look in i3bar (good, prefix icon is present):

![legacy](https://user-images.githubusercontent.com/40846997/83516614-f2128000-a4c6-11ea-8f7e-d40166029618.png)

One iteration of bumblebee-status stdout:

    [
        {
            "background": null,
            "color": "#202020",
            "full_text": "\ue0b2",
            "separator": false,
            "separator_block_width": 0
        },
        {
            "_prefix": "<span font_desc=\"Iosevka 15\" foreground=\"#ff8700\" rise=\"-3500\">\uf318</span>",
            "_raw": "64 14.2",
            "_suffix": "<span font_desc='Iosevka 15' rise='-3500'></span>",
            "align": null,
            "background": "#202020",
            "color": "#7e8aa2",
            "full_text": "<span font_desc=\"Iosevka 15\" foreground=\"#ff8700\" rise=\"-3500\">\uf318</span>64 14.2<span font_desc='Iosevka 15' rise='-3500'></span>",
            "instance": "62b7cc78-a696-41f4-ae59-f1ff33868a0f",
            "markup": "pango",
            "min_width": null,
            "name": "osname",
            "separator": false,
            "separator_block_width": 0
        }
    ]

Setup 2:

revision: 9cadcee

i3 bar config:

    bar {
            id top
            workspace_buttons no
            font pango:Iosevka 9
            position top
            status_command python3 bumblebee-status/bumblebee-status \
                -m spacer:osname \
                -p osname.text="64 14.2" \
                -t mustang-powerline \
                -i nerd-fonts
    }

ported theme in `~/.config/bumblebee-status/themes/mustang-powerline.json`:

    {
        "icons": [ "awesome-fonts" ],
        "defaults": {
            "default-separators": false,
            "separator-block-width": 0,
            "prefix": {
                "pango": {
                    "foreground": "#ff8700",
                    "font_desc": "Iosevka 15",
                    "rise": -3500
                }
            }
        },
        "cycle": [
            {
                "fg": "#7e8aa2",
                "bg": "#202020"
            },
            {
                "fg": "#b1d631",
                "bg": "#2d2d2d"
            }
        ]
    }

How does it look in i3bar (bad, prefix icon is gone):

![master](https://user-images.githubusercontent.com/40846997/83516642-fc347e80-a4c6-11ea-86ed-a4b7a93656aa.png)

One iteration of bumblebee-status stdout:

    [
        {
            "_decorator": true,
            "background": null,
            "color": "#202020",
            "full_text": "\ue0b2",
            "instance": "6ef6fa5b-3622-423a-ab15-95e765fc1d20",
            "name": "osname",
            "separator": false,
            "separator_block_width": 0
        },
        {
            "_prefix": "<span foreground=\"#ff8700\" font_desc=\"Iosevka 15\" rise=\"-3500\"></span>",
            "_raw": "64 14.2",
            "_suffix": "",
            "align": null,
            "background": "#202020",
            "color": "#7e8aa2",
            "full_text": "<span foreground=\"#ff8700\" font_desc=\"Iosevka 15\" rise=\"-3500\"></span>64 14.2",
            "instance": "6ef6fa5b-3622-423a-ab15-95e765fc1d20",
            "markup": "pango",
            "name": "osname",
            "separator": false,
            "separator_block_width": 0
        }
    ]

**Comment by tobi-wan-kenobi** [comment]
Can you please show the contents of the nerd-fonts.json as well?

**Comment by tobi-wan-kenobi** [comment]
For example, I just edited `awesome-fonts.json` to look like this in the defaults (no other changes!):

```
  "defaults": {
    "separator": "",
    "padding": " ",
    "unknown": { "prefix": "" },
    "separator-block-width": 0,
    "default-separators": false,
        "prefix": {
                "pango": {
                        "foreground": "red",
                        "background": "blue",
                        "rise": "-3500"
                }
        }
  },
```

(note the pango stuff)

The - ahem - aesthetically questionable result is:

![uglystuff](https://user-images.githubusercontent.com/6568442/83555663-e64aac00-a50e-11ea-84b9-6a1a5872b479.png)

**Comment by somospocos** [comment]
It's at the top of the bug report.

It's just the minimized version of https://github.com/somospocos/bumblebee-status-contrib/blob/master/themes/icons/nerd-fonts.json , with only "defaults" and "osname" keys. I spotted the problem on the very first big config from the README of that repo and noticed that only the first widget is missing the icon. So I chopped that to a minimum that shows the problem.

**Comment by tobi-wan-kenobi** [comment]
Ah, sorry, I overlooked that.

I still can't reproduce that, weird. I suspect that the iconset configured in the theme is somehow interfering... To confirm that suspicion, could you please try to directly modify the iconset you want changed (i.e. the one the theme is using)?

I don't have a lot of time right now, but I will spend the weekend trying to get to the bottom of this. I think the merging of multiple overlaying/conflicting theme files isn't elegant yet.

Thanks for the report!

**Comment by somospocos** [comment]
I think I have just narrowed it down a bit more.

In setup 2 above (the one that doesn't work), I've got rid of the alias for spacer (I used the spacer module name and spacer.text parameter name directly, for testing purposes). I also modified nerd-fonts.json to use spacer instead of osname. And the icon appeared.

So it looks like when I have a `-m spacer:ALIAS` in bumblebee-config invocation and an `"ALIAS"` key in the icons file - those don't play nice.

**Comment by tobi-wan-kenobi** [comment]
@somospocos can you please let me know whether the commit fixes the issue for you?

I am not sure I will keep that commit in, though. It seems too easy to break stuff (e.g. `-m spacer:disk` would automatically give the disk module themeing).

How about a generic parameter `-m spacer:test -p test.theme=disk` to specify a user-theme? Would that suit your purpose? (Provided the commit above even fixes your issue)

**Comment by somospocos** [comment]
> whether the commit fixes the issue

Looks like it does. After pulling this change the icon is back in both the minimal one-module setup listed at the top of this issue, as well as in the full setup from the contrib repo.

> It seems too easy to break stuff (e.g. -m spacer:disk would automatically give the disk module themeing).

Am I right to assume that this example makes you worried about a possible name clash between custom alias names that an user can choose and names of "built-in" modules? If this is true, then I think this goes past the scope of this particular icon trouble involving the use of spacer module with alias, so shall we have a separate issue about it perhaps?

> How about a generic parameter -m spacer:test -p test.theme=disk to specify a user-theme?

I did not understand this question, but there are two things that I don't understand here, so I'll go into details on them one by one:

1. What do you mean by "user-theme"? I don't have such a definition in my bumblebee-status knowledge base. To me a theme is just a theme. I also went and skimmed one more time through the docs at https://bumblebee-status.readthedocs.io/en/master/development/theme.html before posting this, just to make sure I'm not missing anything, and I can't find anything about "user-themes" there either. Is this some kind of new concept that I need to learn?

2. Regarding `-m spacer:test -p test.theme=disk`. This is an example of using the spacer module, of using it aliased as `test` name, and of setting a `theme` parameter for it. But what is the meaning of `disk` in this example? Perhaps me not understanding this is due to me not understanding 1. above?

**Comment by tobi-wan-kenobi** [comment]
Yeah, OK, so I did not explain that very well. Too many concepts in my head that aren't written down anywhere.

What I meant was: If a user had the option to override the theme to be applied on a per-module basis (`-p test.theme=disk` would mean "to the module called test, apply the 'disk' theme"), would that suit your requirement?

So, in your use case, if I understand it correctly, you would leave the JSON exactly as it is, and when calling `bumblebee-status`, you'd add `-p osname.theme=osname`. That would make the module "osname" (which is just an alias for your particular "spacer") use the entry in the theme that's called "osname".

I think that would be a pretty elegant way around the risk of having modules "accidentially" themed wrongly because of their alias.

**Comment by tobi-wan-kenobi** [comment]
You make valid points, especially:

1. Using <module>.**theme** would be confusing, as it's basically overloading the meaning of the term "theme" - point taken :)
2. Backwards compatibility - it used to work, so why doesn't it now? This puts to rest my concerns about "accidental" theme application. Anyhow, I think that the whole aliasing feature is probably something only used by a small minority of users.

To sum up: You've convinced me that keeping the functionality as it is now is totally fine :) I think that means we can close the ticket, right?

Also: No, I don't find them counterproductive at all, on the contrary, I feel humbled that anyone would take the time and energy necessary to compose their thoughts so thoroughly just for my small project. If there is one frustration then it is that I take time to process this information and there's an asynchronous question/answer/counter-question cycle that could be much easier resolved in a call or a chat. Food for thought there, but TBH, I am not in a phase of my life where I can spend much time hanging around on IRC either :( (those days are sadly past)

So thanks a lot for your thoughtful input that definitely improves bumblebee-status a lot!

**Comment by somospocos** [comment]
TLDR: yes, we can close the ticket if you are happy with the current state.

> Anyhow, I think that the whole aliasing feature is probably something only used by a small minority of users.

But it is still an useful feature regardless of how popular it is, isn't it? Is there another way of using a module multiple times without aliasing it? Think of the very general spacer module that just displays some static text. So I use it once to display some text. What is my option if I want to display some other static text a few modules later, if I wouldn't be able to use it one more time? Hypothetically, write a simple custom third party module that displays some static text, but this would be surely more complex than just using aliases. Or is there another way to achieve the same result that I'm missing?

> To sum up: You've convinced me that keeping the functionality as it is now is totally fine :) I think that means we can close the ticket, right?

Since the change you made solves the raised particular issue, if you're happy with it, then sure - feel free to keep it in master and mark the issue as closed.

> If there is one frustration then it is that I take time to process this information and there's an asynchronous question/answer/counter-question cycle that could be much easier resolved in a call or a chat.

I absolutely have no problem with waiting between posts, as long as it doesn't imply having to wait an indefinite amount of time - as long as I'll know it's one day, one week, one year, I'll just plan my schedule ahead for other things until then, this pretty much eliminates the frustration factor for me. This is a free software project and not having quotas for the amount of code to produce or very tight deadlines for releases is actually a feature, not a bug: everyone can spend as much time as he can and as he wants - this is even more important to avoid getting stressed, I think every free software contributor has the right to get lazy on purpose sometimes :-)

**Comment by tobi-wan-kenobi** [comment]
> But it is still an useful feature regardless of how popular it is, isn't it? Is there another way of using a module multiple times without aliasing it?

Fully agreed - and I intend to keep the aliasing in & supported, definitely. Supporting multiple instances of the same module with different sets of parameters was exactly the usecase the aliases were designed for.

> Since the change you made solves the raised particular issue, if you're happy with it, then sure - feel free to keep it in master and mark the issue as closed.

Yes, I'd say so. It restores backwards compatibility and the diff looks clean to me. If you have any further issues with the functionality or suggestions for improvement, please let me know.

> This is a free software project and not having quotas for the amount of code to produce or very tight deadlines for releases is actually a feature, not a bug: everyone can spend as much time as he can and as he wants - this is even more important to avoid getting stressed, I think every free software contributor has the right to get lazy on purpose sometimes :-)

Amen :) I guess I'm just conditioned to dislike open tickets *a lot* :P 

Thanks a lot for your feedback & have a nice day/evening/whatever, depending on your timezone!
